### PR TITLE
date component example with advance validation

### PR DESCRIPTION
**Pull request by sulthan-ahmed**
* A simple date-component example with a bit of advance validation
* The example is now concise whereas previously it was aimed at a dev that was quite a novice
* Detail information about advance validation

### CODE REVIEW

**Comment by JoeChapman** [comment]
I think in general you should split this PR into two, one for date-component example and one for the amend to the summary page base url

File: apps/component-date/fields/index.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+const dateComponent = require('hof-component-date');
+
+module.exports = {
+  name: 'component-date',
+  'date-of-birth': dateComponent('date-of-birth', {
+    // before means a date cannot be in the future
+    validate: ['required', 'date', 'before']
```
Comment (JoeChapman) [inline apps/component-date/fields/index.js]: Isn't the `date` validator always applied to a date component, does it need to be stated explicitly?

File: apps/component-date/fields/index.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+const dateComponent = require('hof-component-date');
+
+module.exports = {
+  name: 'component-date',
```
Comment (JoeChapman) [inline apps/component-date/fields/index.js]: I don't think this does anything here.

File: apps/component-date/fields/index.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  'date-of-birth': dateComponent('date-of-birth', {
+    // before means a date cannot be in the future
+    validate: ['required', 'date', 'before']
+  }),
+  'passport-expiry': dateComponent('passport-expiry', {
+    // after means a date cannot be in the past
+    validate: ['required', 'date', 'after']
```
Comment (JoeChapman) [inline apps/component-date/fields/index.js]: Isn't the `date` validator always applied to a date component, does it need to be stated explicitly?

File: apps/component-date/index.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+const summary = require('hof-behaviour-summary-page');
+
+module.exports = {
+  name: 'component-date',
+  // this will mean you will need to go to http://localhost:8080/component-date
```
Comment (JoeChapman) [inline apps/component-date/index.js]: Only if your app is running at `http://localhost:8080`. IMO, you should either be clear about the address or not mentions this at all.

File: apps/component-date/index.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  baseUrl: '/component-date',
+  steps: {
+    '/date-examples': {
```
Comment (JoeChapman) [inline apps/component-date/index.js]: perhaps split these over two steps, and don't use the word 'examples' in the paths?

File: apps/component-date/fields/index.js (hunk 2 above)
Comment (sulthan-ahmed) [inline apps/component-date/fields/index.js]: you are correct

File: apps/component-date/fields/index.js (hunk 1 above)
Comment (sulthan-ahmed) [inline apps/component-date/fields/index.js]: as discussed there is no date validator by default.  In fact, you get some weird shizzle when you just put in an invalid date and use before and required
screen shot 2017-06-01 at 14 51 18

File: apps/component-date/fields/index.js (hunk 3 above)
Comment (sulthan-ahmed) [inline apps/component-date/fields/index.js]: as above, it does need to be stated explicitly

File: apps/component-date/README.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Date component example
+============
+
```
Comment (sulthan-ahmed) [inline apps/component-date/README.md]: Done

File: apps/component-date/fields/index.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+const dateComponent = require('hof-component-date');
+
+module.exports = {
+  'date-of-birth': dateComponent('date-of-birth', {
+    // before means a date cannot be in the future
```
Comment (JoeChapman) [inline apps/component-date/fields/index.js]: I'm not sure what happened here, did you mean to have validators?

File: apps/component-date/index.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+const summary = require('hof-behaviour-summary-page');
+
+module.exports = {
+  name: 'component-date',
+  baseUrl: '/component-date',
```
Comment (JoeChapman) [inline apps/component-date/index.js]: does HOF use the `name` as the `baseUrl` or vice versa? I think only one of them is required when they are the same

File: apps/component-date/index.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  steps: {
+    '/date-of-birth': {
+      fields: ['date-of-birth'],
```
Comment (JoeChapman) [inline apps/component-date/index.js]: Can't quite remember but I don't think single field names need be declared in an array, they get cast to arrays when processed.

File: apps/component-date/index.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+      next: '/passport-expiry'
+    },
+    '/passport-expiry': {
+      fields: ['passport-expiry'],
+      next: '/confirm'
```
Comment (JoeChapman) [inline apps/component-date/index.js]: When the steps are ordered in the ordered sequentially like this do you need to declare the `next` step or is it implicit?

File: apps/component-date/README.md (hunk 6 above)
Comment (lennym) [inline apps/component-date/README.md]: The intention (or at least *my* intention) for these examples is that they would provide tutorial-style content for how a new hof user would go about implementing the included functionality.

These would then be dynamically loaded into the `hof-guide`.

Can you write something like that for this example?

File: apps/component-date/index.js (hunk 8 above)
Comment (lennym) [inline apps/component-date/index.js]: Not as far as I know.

Comment (lennym) [inline apps/component-date/index.js]: Can we call it `date-component` since that's the name of the thing? Yoda we are not.

File: apps/component-date/index.js (hunk 10 above)
Comment (lennym) [inline apps/component-date/index.js]: It is not implicit.

File: apps/component-date/translations/src/en/fields.json [hunk 11]
```diff
@@ -0,0 +1,10 @@
+{
+  "date-of-birth": {
+    "legend": "Date of birth",
+    "hint": "For example, 12 3 1980"
+  },
+  "passport-expiry": {
+    "legend": "Passport expiry date",
+    "hint": "For example, 30 04 2028"
```
Comment (lennym) [inline apps/component-date/translations/src/en/fields.json]: That's not a valid passport expiry date. Passports only last 10 years. I also don't care about this, I don't even know why I'm still typing.

File: apps/component-date/fields/index.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  }),
+  'passport-expiry': dateComponent('passport-expiry', {
+    // after means a date cannot be in the past
+    validate: ['required', 'after']
```
Comment (lennym) [inline apps/component-date/fields/index.js]: It'd be good to include an example of `before`/`after` validation with an offset parameter. Mostly because I can't even remember myself how that works.

**Review by lennym — CHANGES_REQUESTED** [review]

File: apps/component-date/index.js (hunk 8 above)
Comment (sulthan-ahmed) [inline apps/component-date/index.js]: I tried to remove one or the other and unfortunately it does need both items even if they have the same name, do you want to change that in hof?

File: apps/component-date/fields/index.js (hunk 12 above)
Comment (JoeChapman) [inline apps/component-date/fields/index.js]: Like this?

```
{
        type: 'before',
        arguments: '2017-06-01T15:56:49.770Z'
}
```

File: apps/component-date/index.js (hunk 9 above)
Comment (sulthan-ahmed) [inline apps/component-date/index.js]: Looks like they do
`TypeError: (options.fields || []).reduce is not a function
    at Wizard._.each (/Users/Sulthan/Development/projects/HO/hof/node_modules/hof-form-wizard/lib/wizard.js:63:45)`
Do you want me to change it in hof?

File: apps/component-date/index.js (hunk 10 above)
Comment (sulthan-ahmed) [inline apps/component-date/index.js]: seems like you do otherwise it doesn't let you go anywhere

File: apps/component-date/index.js (hunk 9 above)
Comment (lennym) [inline apps/component-date/index.js]: It should not change in hof. 

`name` is *required* and needs to match the directory name. `baseUrl` is optional and should default to `/` since *most* apps will be a single app mounted at the domain root.

File: apps/component-date/index.js (hunk 8 above)
Comment (lennym) [inline apps/component-date/index.js]: It should not change in hof.

name is required and needs to match the directory name. baseUrl is optional and should default to / since most apps will be a single app mounted at the domain root.

File: apps/component-date/fields/index.js (hunk 12 above)
Comment (lennym) [inline apps/component-date/fields/index.js]: Ugh. I just looked, and it seems that `before` and `after` take different arguments, which sucks spectacular balls. 

https://github.com/UKHomeOfficeForms/hof-form-controller/blob/master/lib/validation/validators.js#L88-L118

File: apps/component-date/README.md (hunk 6 above)
Comment (sulthan-ahmed) [inline apps/component-date/README.md]: I totally agree with the intention.  I was just planning to put it in a separate PR but probably makes sense to do it now otherwise it might not happen

File: apps/component-date/translations/src/en/fields.json (hunk 11 above)
Comment (sulthan-ahmed) [inline apps/component-date/translations/src/en/fields.json]: LOL
I'll change it

File: apps/component-date/README.md (hunk 6 above)
Comment (lennym) [inline apps/component-date/README.md]: Please add it now. We should consider "go and look at the code" as a failure of documentation, even if we do have nicely commented examples. The code here should be a reference for a finished version, but the README should tell the user how they would implement in their own app.

This also helps to make it clear to novice users what parts of the code example are relevant to the functionality in question. Otherwise they find themselves copying random unrelated things in an effort to "make it look like the example".

Comment (sulthan-ahmed) [inline apps/component-date/README.md]: fair 👍 I'll do that now

File: apps/component-date/index.js (hunk 8 above)
Comment (sulthan-ahmed) [inline apps/component-date/index.js]: @lennym I started calling it `date-component` and searched for it in https://github.com/UKHomeOfficeForms until I remembered we called it `hof-component-date`.  I kept it `component-date` just for consistency.  I can change it if you like but it's a little confusing for a novice that in one place we are saying `date-component` and on the repo we say `hof-component-date`.  However, I do prefer the name date-component

Also, I do sometimes think of you as Yoda

Comment (lennym) [inline apps/component-date/index.js]: Ack, you are right. However I think that's more to do with the repo/module naming convention we have of `hof-<module-type>-<name>` than anything else. I still think `date-component` reads better and is more user friendly.

Also, if you're going to think of me as Yoda then I'm going to have to insist on being able to claim piggy-back rides.

File: apps/component-date/fields/index.js (hunk 3 above)
Comment (sulthan-ahmed) [inline apps/component-date/fields/index.js]: I shall now amend this to reflect the fact that the `date-component` has this by default

File: apps/component-date/index.js (hunk 8 above)
Comment (sulthan-ahmed) [inline apps/component-date/index.js]: fair enough I shall change it 
piggyback? duuude, have you seen the size of me compared to you?

alt text

**Comment by JoeChapman** [comment]
@sulthan-ahmed is this still a work in progress or do you want this PR merged?

**Comment by sulthan-ahmed** [comment]
I was hoping someone could have a look to see why the images aren't showing @JoeChapman and whether I've put the images in the right place would you be able to help?

**Comment by JoeChapman** [comment]
@sulthan-ahmed where are you expecting the images to appear, are they loaded with CSS, or inline HTML?

**Comment by sulthan-ahmed** [comment]
I'm just pointing to the images in the README @JoeChapman via markdown

File: apps/date-component/README.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      next: '/complete'
+    },
+    '/complete': {
+      template: 'confirmation'
+    }
+  }
+```
+Now start your app by running `npm start`. Open http://localhost:8080/date-component in a browser then you should see your app. It should look something like this:
+
+![Date component start page](../../images/date-component-getting-started.png?raw=true)
```
Comment (lennym) [inline apps/date-component/README.md]: Should be `get-started.png`

File: apps/date-component/README.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+module.exports = {
+  'date-of-birth': dateComponent('date-of-birth', {
+    validate: ['required', 'before']
+  })
+};
+```
+
+Now start your app by running `npm start`. Open http://localhost:8080/date-component in a browser then you should see your app using the date-component. It should look something like this:
+
+![Date component field](../assets/images/date-component-date-of-birth.png)
```
Comment (lennym) [inline apps/date-component/README.md]: Image path is incorrect.

File: apps/date-component/README.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+This tutorial explains the steps to create a date using the [hof-component-date](https://github.com/UKHomeOfficeForms/hof-component-date)
+
+## Getting started
+
+Create a directory called date-component and use the [hof-generator](https://github.com/UKHomeOfficeForms/hof-guide/blob/master/source/documentation/getting-started.md) to create your app
+
+Currently the app loads to a name page with a field `input[type=text]`.  We will remove this and create a step that displays a date of birth.
+
+Go into the following directory `date-component-> apps -> date-component` and open the `index.js` file. Rename the first step key from `/name` to `/date-of-birth`.
```
Comment (lennym) [inline apps/date-component/README.md]: I wouldn't worry too much about describing things that a user should already know how to do. i.e. generating an app and adding a field.

The starting point should be "you have an app, and have followed the basic documentation so know how to do standard things".

**Review by lennym — CHANGES_REQUESTED** [review]

**Comment by lennym** [comment]
Why is there a README.html *and* a README.md?

**Comment by lennym** [comment]
It would be good to describe and give examples of how to use some of the more advanced date validation.

**Comment by sulthan-ahmed** [comment]
> Why is there a README.html and a README.md?

:see_no_evil: I haven't got a clue, it could be the add on I used to preview my markdown in sublime.  I'll remove

**Comment by JoeChapman** [comment]
@sulthan-ahmed this needs moar rebase

**Comment by sulthan-ahmed** [comment]
thanks @JoeChapman 

I'll rebase locally.  I'm just working on adding a bit more to this PR around advanced date validation.  Once I've done that then I'll get rid of the WIP from the PR and any commits and push all the changes

File: apps/date-component/README.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
+# Date component example
+
+This tutorial explains the steps to create a date using the [hof-component-date](https://github.com/UKHomeOfficeForms/hof-component-date)
+
+## Getting started
+
+Create a hof app called date-component
```
Comment (JoeChapman) [inline apps/date-component/README.md]: I know we decided not to detail the universe in every tutorial and assume the user already has some experience, but I think I would prefer it if we made it clearer, something like:


-------------------------------------------------------------
### Create the app

```bash
$ hof app date-component 
```

### Edit the app

Open the `./apps/date-component/index.js` and rename the first step from `/name` to `/date-of-birth`. 
And edit the following ....

---------------------------------------------------------------
I think it's important to phrase the instructions succinctly, perhaps as numbered steps, but definitely with order and language that leads into the next instruction. Just read it back to yourself and edit it until it makes sense

Comment (lennym) [inline apps/date-component/README.md]: Remember that these are tutorials for a user to add these pieces of functionality to *their own apps*, not just to recreate the example.

Code examples are useful, but they're not the goal here.

File: apps/date-component/README.md [hunk 17]
```diff
@@ -0,0 +1,82 @@
+# Date component example
+
+This tutorial explains the steps to create a date using the [hof-component-date](https://github.com/UKHomeOfficeForms/hof-component-date). There is also a further explanation of useful date validations.
+
+## Getting started
+
+Add a step that displays a date of birth and a field called date-of-birth.
+
```
Comment (JoeChapman) [inline apps/date-component/README.md]: Maybe just `Add a step that displays a date of birth.`

File: apps/date-component/README.md [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+  steps: {
+    '/date-of-birth': {
+      fields: ['date-of-birth'],
+      next: '/confirm'
+    }
+  }
+```
+
+## Adding the Date-component field
+
```
Comment (JoeChapman) [inline apps/date-component/README.md]: Ehen referring to the dat component, probably best to either be precise `hof-component-date` or verbose as in 'Date Component'

File: apps/date-component/README.md [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+Install the [hof-component-date](https://www.npmjs.com/package/hof-component-date) module
+
+Next, go into the fields directory and open the `index.js` file. Add the field with the key `date-of-birth`. Give this key a property of a `dateComponent()`. The date-component is a function that takes arguments.  Give it the arguments of the name of the field `date-of-birth` and an object validate key with an array of properties `['required']`
+
```
Comment (JoeChapman) [inline apps/date-component/README.md]: Again, I'm not sure what the `date-component` is. sorry to be pedantic. But I think it's important to use consistent and exact language. It should be either `hof-component-date` or more casually the `Date component`. It's either the noun or an adjective. But `date-component` looks technically precise but it not, which I think could be confusing

File: apps/date-component/README.md [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+The above example would mean a user must enter a value before today's date otherwise a validation will be triggered.
+
+### Date argument
+
+A single date argument can be used.  The validator would use the value entered by the user to compare with the date supplied, in the argument.
+
+```
+validate: [{type: 'after', arguments: '1892-01-01'}]
+```
```
Comment (JoeChapman) [inline apps/date-component/README.md]: ooooh, cool that this is documented. Could we make a direct comparison? By which I mean can you write the example with arguments with a type value of `before`?

Comment (JoeChapman) [inline apps/date-component/README.md]: oh, you've done it lower down.

**Comment by JoeChapman** [comment]
Please can you add a link from the main README to this link? Hopefully we'll end up with a list of examples on the front page.

File: apps/date-component/README.md [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+      next: '/confirm'
+    }
+  }
+```
+
+## Adding the Date-component field
+
+Install the [hof-component-date](https://www.npmjs.com/package/hof-component-date) module
+
+Next, go into the fields directory and open the `index.js` file. Add the field with the key `date-of-birth`. Give this key a property of a `dateComponent()`. The date-component is a function that takes arguments.  Give it the arguments of the name of the field `date-of-birth` and an object validate key with an array of properties `['required']`
```
Comment (lennym) [inline apps/date-component/README.md]: I don't think "Give this key a property of a `dateComponent()`" really makes sense. And "is a function that takes arguments" is something of a tautology.

I'd replace this whole paragraph with:

> In your field configuration file, create an instance of a data component. You can do this by calling the data component function with a field name and the options for the field. In this case, our field name is `date-of-birth` and in our options we set a `required` validator.

> Then assign this instance to the appropriate key on your fields config.

File: apps/date-component/README.md [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+### No arguments
+
+If no arguments are supplied then the validator uses the value of the input of the user to compare with today's date.
+
+```
+validate: ['required', 'before']
+```
+
+The above example would mean a user must enter a value before today's date otherwise a validation will be triggered.
```
Comment (lennym) [inline apps/date-component/README.md]: > a validation error will be thrown

File: apps/date-component/README.md [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+
+### Date argument
+
+A single date argument can be used.  The validator would use the value entered by the user to compare with the date supplied, in the argument.
+
+```
+validate: [{type: 'after', arguments: '1892-01-01'}]
+```
+
+This would be mean a user must enter a value after `1892-01-1`. Otherwise a validation will be triggered.
```
Comment (lennym) [inline apps/date-component/README.md]: > `1892-01-01` (missing `0`)

> a validation error will be thrown

File: apps/date-component/fields/index.js [hunk 24]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+const dateComponent = require('hof-component-date');
+
+module.exports = {
+  'date-of-birth': dateComponent('date-of-birth', {
+    // before means a date cannot be in the future
+    validate: ['required',
+    {type: 'before', arguments: [18, 'years', 11, 'months']},
+    {type: 'after', arguments: '1892-01-01'}]
```
Comment (lennym) [inline apps/date-component/fields/index.js]: Can we format this as follows:

```js
validate: [
  'required',
  { ... },
  { ... }
]
```

Thanks.

**Review by lennym — CHANGES_REQUESTED** [review]

File: apps/date-component/README.md (hunk 17 above)
Comment (sulthan-ahmed) [inline apps/date-component/README.md]: Fair

File: apps/date-component/README.md (hunk 19 above)
Comment (sulthan-ahmed) [inline apps/date-component/README.md]: I agree totally

File: apps/date-component/README.md (hunk 21 above)
Comment (sulthan-ahmed) [inline apps/date-component/README.md]: That sounds much better thanks

**Comment by sulthan-ahmed** [comment]
Great great comments, thanks @lennym and @JoeChapman

File: apps/date-component/README.md [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+      next: '/confirm'
+    }
+  }
+```
+
+## Adding the Date component field
+
+Install the [hof-component-date](https://www.npmjs.com/package/hof-component-date) module
+
+In your field configuration file, create an instance of a data component. You can do this by calling the data component function with a field name and the options for the field. In this case, our field name is date-of-birth and in our options we set a required validator.
```
Comment (JoeChapman) [inline apps/date-component/README.md]: `data component` <- Don't copy Lenny

Comment (JoeChapman) [inline apps/date-component/README.md]: Also, an instance of a "date component"? It would be an 'instance of `hof-component-date`'

Comment (lennym) [inline apps/date-component/README.md]: You said it could either be referred to as a "date component" (casually - no hyphen) or "hof-component-date". What's your beef?

Comment (lennym) [inline apps/date-component/README.md]: Do copy me, unless I'm wrong.

Comment (JoeChapman) [inline apps/date-component/README.md]: > In your field configuration file, create an instance of `hof-component-date`. You can do this by calling the date component function with a field name and the options for the field. In this case, our field name is `date-of-birth` and in our options we set a `required` validator.
